Various improvements to HPOS settings

jorgeatorres opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->
This PR makes a few changes to the HPOS settings screen (and some related functions). This is based on feedback we've received, but also things we've been notices (such as duplicate queries, unused functions, etc.).

Some of the changes contained in this PR:

- In 902e1ec2f6f2d2cb99629ddf71a509cf7d60321c, we remove a hard-coded style for settings and make it a CSS class instead, which is more flexible and also plays better with colors defined elsewhere. This might seem unrelated to HPOS, but was added specifically in the context of HPOS settings, so seems reasonable to update that here too.
- Removal of an internal sync option (`woocommerce_initial_orders_pending_sync_count`) that is not used at all in the sync code. This allows us to deprecate `DataSynchronizer::get_sync_status()`, which at this point, is nothing but an unnecessary wrapper over methods counting the number of orders pending sync.
- When generating settings/labels for HPOS, we now use the "cached" version of the orders pending sync count function, which removes query duplication on this screen.
- Some UI changes such as removal of an unnecessary tooltip, and proper formatting for numbers. Also, screens were tweaked as described below:

   | Before | Now | Why? |
   |--------|--------|--------|
   | <img width="1066" alt="Screenshot 2024-05-13 at 11 27 03" src="https://github.com/woocommerce/woocommerce/assets/184724/32aa1fd0-b83b-4b79-b7ca-bd91c1c79e8b"> | <img width="1108" alt="Screenshot 2024-05-13 at 11 27 10" src="https://github.com/woocommerce/woocommerce/assets/184724/177db290-4f81-4680-aa46-c60ed1c616df"> | The "pending sync" wording has been reported (several times) as confusing users, as even with compatibility mode disabled, it seems to imply you _have_ to sync orders. I've changed this for "out of sync" and moved the bit about changing datastores first, to make it clear that it's only when changing datastores that this is relevant. |
   | <img width="721" alt="Screenshot 2024-05-13 at 11 27 24" src="https://github.com/woocommerce/woocommerce/assets/184724/64424749-1ac7-41d5-93a5-587233334085"> | <img width="717" alt="Screenshot 2024-05-13 at 11 27 31" src="https://github.com/woocommerce/woocommerce/assets/184724/ffc9ae8d-970a-407d-87d5-aac92003cad5"> | It's now possible to stop the manual sync, which means merchants no longer have to wait until it completes if they clicked it by mistake. |

Closes #43268.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure your site has a few orders or use Smooth Generator (if needed).
2. Go to WC > Settings > Advanced > Features and make sure HPOS is selected as order data storage and that compatibility mode is _disabled_.
   If necessary, use `wp wc hpos sync` to first sync orders so that all options become available.
3. **Check that there's no synchronization query duplication.**
   1. Install and activate [Query Monitor](https://wordpress.org/plugins/query-monitor/) to be able to detect duplicate queries.
   2. Go to WC > Settings > Advanced > Features and make sure that Query Monitor doesn't report any duplication of queries related to `Automattic\WooCommerce\Internal\DataStores\Orders\DataSynchronizer`.
   3. (Optional) Contrast this with `trunk` where duplication is reported:
      <img width="1331" alt="Screenshot 2024-05-13 at 11 29 58" src="https://github.com/woocommerce/woocommerce/assets/184724/1c0aeff0-05ab-4e42-b96b-bc185b97e275">
4. Run `wp wc hpos cleanup all` so that order storage is no longer in sync.
5. Go to WC > Settings > Advanced > Features and confirm that you see a message about orders being out of sync.
6. **Check that sync can be manually started and stopped.**
   i. On WC > Settings > Advanced > Features click "Sync orders now" to start a manual order sync.
   ii. Confirm that after the page refreshes, you see a message about orders being synced and there's a link to "Stop sync".
   iii. Refresh the page until you see number of unsynced orders go down (might take a bit depending on when A-S runs).
   iv. Quickly click "Stop sync".
   v. Refresh the page and confirm that orders have stopped syncing and that "Sync orders now" appears once again.
7. Run `wp wc hpos cleanup all` again to make sure the datastores are not in sync.
8. Enable compatibility mode on the WC > Settings > Advanced > Features screen, and confirm that orders are syncing and that you can't choose a different data storage until the process is complete.

#### Test `description_is_error`

This PR also introduces a change in how settings with `description_is_error` are treated. Instead of hardcoding the styles in the HTML tags, we now use a CSS class that uses the color defined via SCSS.

1. Build WC so that legacy assets are properly built and linked.
2. Make sure datastores are not in sync. Run `wp wc hpos cleanup all` if necessary.
3. Go to WC > Settings > Advanced > Features and you should see the following message inside the HPOS section:
   <img width="884" alt="Screenshot 2024-05-20 at 09 44 01" src="https://github.com/woocommerce/woocommerce/assets/184724/21411681-32f3-4b8a-b3f2-899b73ebc013">
4. Activate the following snippet on your site:
   ```php
   add_filter( 'wc_allow_changing_orders_storage_while_sync_is_pending', '__return_true' );
   ````
5. Go back to WC > Settings > Advanced > Features and the message from step 3, should've changed to the following and be displayed in red:
   <img width="1003" alt="Screenshot 2024-05-20 at 09 46 45" src="https://github.com/woocommerce/woocommerce/assets/184724/64ac3d0e-6ca6-4b3a-a352-81f99943bddb">


<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

jorgeatorres committed:
Properly format order counts

jorgeatorres committed:
Remove superfluous tooltip

jorgeatorres committed:
Add "is-error" class for admin settings

jorgeatorres committed:
Remove unused option in DataSynchronizer

jorgeatorres committed:
Use non-persistent group "counts" for caching count of orders pending sync

jorgeatorres committed:
Drop redundant method from COT controller

jorgeatorres committed:
Standardize usage of <strong> in COT settings

jorgeatorres committed:
Drop unused import

jorgeatorres committed:
Allow manual sync to be stopped

jorgeatorres committed:
Reword messages in HPOS screen

jorgeatorres committed:
Make PHPCS happy

jorgeatorres committed:
Add changelog

jorgeatorres committed:
Fix unit test

jorgeatorres committed:
Fully deprecated DataSynchronier::get_sync_status()

naman03malhotra commented on plugins/woocommerce/includes/admin/class-wc-admin-settings.php:
```diff
@@ -805,19 +805,18 @@ public static function get_field_description( $value ) {
-			$extra_description_style = $description_is_error ? " style='color:red'" : '';
+			$error_class = ( ! empty( $value['description_is_error'] ) ) ? 'is-error' : '';
 
 			if ( $description && in_array( $value['type'], array( 'textarea', 'radio' ), true ) ) {
 				$description = '<p style="margin-top:0">' . wp_kses_post( $description ) . '</p>';
 			} elseif ( $description && in_array( $value['type'], array( 'checkbox' ), true ) ) {
 				$description = wp_kses_post( $description );
 			} elseif ( $description ) {
-				$description = '<p class="description"' . $extra_description_style . '>' . wp_kses_post( $description ) . '</p>';
+				$description = '<p class="description ' . $error_class . '">' . wp_kses_post( $description ) . '</p>';
 			}
 
 			if ( $tooltip_html && in_array( $value['type'], array( 'checkbox' ), true ) ) {
-				$tooltip_html = '<p class="description"' . $extra_description_style . '>' . $tooltip_html . '</p>';
+				$tooltip_html = '<p class="description ' . $error_class . '">' . $tooltip_html . '</p>';
```
Great work cleaning this up. Can you please clarify the instructions on how to test this change? It would also be helpful when Solaris or our external partners tests this PR.

jorgeatorres replied:
I've updated the testing instructions with a section on this specific change.

naman03malhotra commented on plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:
```diff
@@ -604,10 +620,28 @@ private function get_hpos_setting_for_sync() {
+					number_format_i18n( $orders_pending_sync_count )
 				);
+
+				if ( ! $sync_enabled ) {
```
I might have the context missing here, what does `sync_enabled` imply? For me it seems counter-intuitive to render stop sync button when `sync_enabled` is `false`

jorgeatorres replied:
`$sync_enabled = true` means compatibility mode is enabled. This mode takes care of both the initial sync of orders but also propagating changes immediately when orders are updated, so there's no need to start or stop anything.

The "sync now" link allows you to trigger a one-off sync without necessarily enabling compatibility mode. The "stop" button would allow you to cancel that one-off sync while it is in progress. This is particularly important when you have a ton of orders, as this can take a while and we currently don't offer a way to stop this process (unless you want to dig into A-S jobs).

naman03malhotra commented on plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:
```diff
@@ -586,14 +602,14 @@ private function get_hpos_setting_for_sync() {
 						_n(
-							"There's %d order pending sync. <b>Switching data storage while sync is incomplete is dangerous and can lead to order data corruption or loss!</b>",
-							'There are %d orders pending sync. <b>Switching data storage while sync is incomplete is dangerous and can lead to order data corruption or loss!</b>',
+							"There's %s order pending sync. <strong>Switching data storage while sync is incomplete is dangerous and can lead to order data corruption or loss!</strong>",
+							'There are %s orders pending sync. <strong>Switching data storage while sync is incomplete is dangerous and can lead to order data corruption or loss!</strong>',
```
Maybewe can extract the common message out, wdyt? For a moment, I was confused as to why we had two similar messages, and then I realized.

Seems duplicate, can be extracted out.
```
<strong>Switching data storage while sync is incomplete is dangerous and can lead to order data corruption or loss!</strong>
```

jorgeatorres replied:
I guess there might be some i18n issues with separating the strings, but I did it anyways in d55c61cd36668f6a787b88ac3392809595bc0082. We can reassess later if this ever becomes a problem for translations, which it most likely won't.

naman03malhotra reviewed (commented):
Thanks for all these awesome imporvements to HPOS! 🎉 

I did one pass of review and overall it looks great, but I am yet to test a few minor things which I will take up next week.

Overall, I am super glad to see this PR, thank you for working on it 🙇.
![image](https://github.com/woocommerce/woocommerce/assets/15019298/10e37e28-20ef-42e9-b792-9a4f460e462c)
![image](https://github.com/woocommerce/woocommerce/assets/15019298/ac15f909-7e18-4efe-abc0-626e4e3185a5)

naman03malhotra reviewed (commented):
Out of context of this PR probably, the sync status doesn't get updates unless the page is fully reloaded. Probably not even needed, but good to have enhancement if we can update the status as well either by polling or websocket.

![image](https://github.com/woocommerce/woocommerce/assets/15019298/67f37afe-60c4-41ab-afeb-5e3eb1c217dd)

jorgeatorres committed:
Extract translatable string to prevent duplication

jorgeatorres commented:
Hey @naman03malhotra! Thanks for the review. I've responded / addressed your comments. Please let me know what you think!

> Out of context of this PR probably, the sync status doesn't get updates unless the page is fully reloaded. Probably not even needed, but good to have enhancement if we can update the status as well either by polling or websocket.

I agree that this isn't the ideal UI, but this would require some changes that are probably not worth the effort right now for a screen that is seldom used. Polling (or similar) to get an updated count is not necessarily the most problematic aspect (though it does involve some heavy queries), but we would also need to properly update all the UI depending on the result (which is much more involved as you see there are a ton of different messages that can appear and interactions between the different settings). There's also the issue of nonce invalidation, etc.

naman03malhotra commented on plugins/woocommerce/client/legacy/css/admin.scss:
```diff
@@ -5133,6 +5133,10 @@ img.help_tip {
 			p.description {
 				margin-bottom: 8px;
+
+				&.is-error, span.is-error {
```
I don't see a span tag heere, maybe just `&.is-error` is enough?
![image](https://github.com/woocommerce/woocommerce/assets/15019298/5624ddc1-86e8-44b1-a257-35c69b3de567)

jorgeatorres replied:
Yep. The idea is that sometimes we might not want the whole description to be considered "error". This would allow you to add some description text and highlight the "error" part by enclosing it in a `<span>` with the `is-error` class.

naman03malhotra replied:
But `is-error` is currently applied on `<p>` tag, do you mean for the future use-case we are adding span here?

naman03malhotra replied:
Or maybe you meant, `p.is-error`?

jorgeatorres replied:
Yes, `is-error` is currently applied on the `p.description` tag if you want the full message to be error colored, but what if you want the description to render normally except for a specific part.

Say, if you create a setting with a description like `This is my description. <span class="is-error">This I want to highlight as an error.</span> But then the rest is normal.`

The (S)CSS above would let you do that. You could have the full `p` be colored as an error or just a `span` inside the description.

naman03malhotra replied:
Oh, got it! thanks for explaining.

naman03malhotra commented on plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:
```diff
@@ -622,39 +655,30 @@ private function get_hpos_setting_for_sync() {
 							_n(
-								"There's %d order pending sync. You can switch order data storage <strong>only when the posts and orders tables are in sync</strong>.",
-								'There are %d orders pending sync. You can switch order data storage <strong>only when the posts and orders tables are in sync</strong>.',
+								"You can switch order data storage <strong>only when the posts and orders tables are in sync</strong>. There's currently %s order out of sync.",
+								'You can switch order data storage <strong>only when the posts and orders tables are in sync</strong>. There are currently %s orders out of sync. ',
```
>You can switch order data storage <strong>only when the posts and orders tables are in sync</strong>.

We can extract out the common message here too, but I'll leave it upto you to do it part of this PR or another one.

naman03malhotra commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -459,7 +463,7 @@ public function get_current_orders_pending_sync_count( $use_cache = false ): int
 
 		if ( $use_cache ) {
-			$pending_count = wp_cache_get( 'woocommerce_hpos_pending_sync_count' );
+			$pending_count = wp_cache_get( 'woocommerce_hpos_pending_sync_count', 'counts' );
```
Nicee 🙌

naman03malhotra reviewed (approved):
Thanks for making all the imporant changes as part of this PR. I've tested and it works well. I've left some comments none of which are blocking for this PR to be merged.

zaderade commented:
I don't think the deeper source of confusion has been resolved. People are confused because they are told that posts and orders are out of sync again a few days _after_ successfully enabling HPOS and disabling compatibility mode. After enabling HPOS, the out of sync message is only relevant if a user desires compatibility mode, so it should otherwise be hidden. I flesh this out at https://wordpress.org/support/topic/hpos-sync-x-pending-orders/page/2/#post-17826401 and suggest programming logic to disambiguate.
